MariaDB not working with Index over VARCHAR(255) and TEXT

Hi,
Keycloak Phase2 Image (quay.io/phasetwo/phasetwo-keycloak:26.5.7) has a problem to create index IDX_ORG_ATTR_NAME_VALUE because of column types, no matter if it is MariaDB 10.11 / MariaDB 11.8 (see stacktrace attached):

2026-04-28 15:38:40,141 ERROR [liquibase.changelog.ChangeSet] (main) ChangeSet META-INF/jpa-changelog-phasetwo-20260421.xml::add-organization-attribute-name-value-index::phasetwo encountered an exception.: liquibase.exception.DatabaseException: (conn=2193) Specified key was too long; max key length is 3072 bytes [Failed SQL: (1071) CREATE INDEX IDX_ORG_ATTR_NAME_VALUE ON keycloak.ORGANIZATION_ATTRIBUTE(NAME, VALUE)]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:520)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:85)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:188)
	at liquibase.executor.AbstractExecutor.execute(AbstractExecutor.java:148)
	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1198)
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:816)
	at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(UpdateVisitor.java:127)
	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:71)
	at liquibase.changelog.ChangeLogIterator.lambda$run$0(ChangeLogIterator.java:133)
	at liquibase.Scope.lambda$child$0(Scope.java:216)
	at liquibase.Scope.child(Scope.java:225)
	at liquibase.Scope.child(Scope.java:215)
	at liquibase.Scope.child(Scope.java:194)
	at liquibase.changelog.ChangeLogIterator.lambda$run$1(ChangeLogIterator.java:122)
	at liquibase.Scope.lambda$child$0(Scope.java:216)
	at liquibase.Scope.child(Scope.java:225)
	at liquibase.Scope.child(Scope.java:215)
	at liquibase.Scope.child(Scope.java:194)
	at liquibase.Scope.child(Scope.java:282)
	at liquibase.Scope.child(Scope.java:286)
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:91)
	at liquibase.command.core.AbstractUpdateCommandStep.lambda$run$0(AbstractUpdateCommandStep.java:114)
	at liquibase.Scope.lambda$child$0(Scope.java:216)
	at liquibase.Scope.child(Scope.java:225)
	at liquibase.Scope.child(Scope.java:215)
	at liquibase.Scope.child(Scope.java:194)
	at liquibase.command.core.AbstractUpdateCommandStep.run(AbstractUpdateCommandStep.java:112)
	at liquibase.command.core.UpdateCommandStep.run(UpdateCommandStep.java:100)
	at liquibase.command.CommandScope.lambda$execute$6(CommandScope.java:263)
	at liquibase.Scope.child(Scope.java:225)
	at liquibase.Scope.child(Scope.java:201)
	at liquibase.command.CommandScope.execute(CommandScope.java:251)
	at liquibase.Liquibase.lambda$update$0(Liquibase.java:216)
	at liquibase.Scope.lambda$child$0(Scope.java:216)
	at liquibase.Scope.child(Scope.java:225)
	at liquibase.Scope.child(Scope.java:215)
	at liquibase.Scope.child(Scope.java:194)
	at liquibase.Liquibase.runInScope(Liquibase.java:1366)
	at liquibase.Liquibase.update(Liquibase.java:205)
	at liquibase.Liquibase.update(Liquibase.java:188)
	at liquibase.Liquibase.update(Liquibase.java:175)
	at org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.updateChangeSet(QuarkusJpaUpdaterProvider.java:191)
	at org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:117)
	at org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:84)
	at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.update(QuarkusJpaConnectionProviderFactory.java:285)
	at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.createOrUpdateSchema(QuarkusJpaConnectionProviderFactory.java:267)
	at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.postInit(QuarkusJpaConnectionProviderFactory.java:124)
	at org.keycloak.services.DefaultKeycloakSessionFactory.initializeProviders(DefaultKeycloakSessionFactory.java:168)
	at org.keycloak.services.DefaultKeycloakSessionFactory.initializeProviders(DefaultKeycloakSessionFactory.java:165)
	at org.keycloak.services.DefaultKeycloakSessionFactory.initProviderFactories(DefaultKeycloakSessionFactory.java:142)
	at org.keycloak.services.DefaultKeycloakSessionFactory.initProviderFactories(DefaultKeycloakSessionFactory.java:126)
	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:87)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:61)
	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:76)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.onStartupEvent(QuarkusKeycloakApplication.java:51)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication_Observer_onStartupEvent_VUhusbW4xizn9A5dimKJUcfvjHs.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:365)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:347)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:81)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:163)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:114)
	at io.quarkus.runner.recorded.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)
	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:127)
	at org.keycloak.quarkus.runtime.cli.Picocli.start(Picocli.java:912)
	at org.keycloak.quarkus.runtime.cli.command.AbstractAutoBuildCommand.runCommand(AbstractAutoBuildCommand.java:139)
	at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.lambda$call$1(AbstractCommand.java:66)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.call(AbstractCommand.java:65)
	at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.call(AbstractCommand.java:34)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2031)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:180)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:102)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:79)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:69)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:37)
Caused by: java.sql.SQLSyntaxErrorException: (conn=2193) Specified key was too long; max key length is 3072 bytes
	at org.mariadb.jdbc.export.ExceptionFactory.createException(ExceptionFactory.java:292)
	at org.mariadb.jdbc.export.ExceptionFactory.create(ExceptionFactory.java:386)
	at org.mariadb.jdbc.message.ClientMessage.readPacket(ClientMessage.java:187)
	at org.mariadb.jdbc.client.impl.StandardClient.readPacket(StandardClient.java:1376)
	at org.mariadb.jdbc.client.impl.StandardClient.readResults(StandardClient.java:1315)
	at org.mariadb.jdbc.client.impl.StandardClient.readResponse(StandardClient.java:1234)
	at org.mariadb.jdbc.client.impl.StandardClient.execute(StandardClient.java:1158)
	at org.mariadb.jdbc.Statement.executeInternal(Statement.java:1003)
	at org.mariadb.jdbc.Statement.execute(Statement.java:1132)
	at org.mariadb.jdbc.Statement.execute(Statement.java:471)
	at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:222)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:512)
	... 86 more

2026-04-28 15:38:40,155 INFO  [com.arjuna.ats.jbossatx] (main) ARJUNA032014: Stopping transaction recovery manager

Best regards
Alex

--- a/src/main/resources/META-INF/jpa-changelog-phasetwo-20260421.xml
+++ b/src/main/resources/META-INF/jpa-changelog-phasetwo-20260421.xml
@@ -7,10 +7,10 @@
                 <indexExists indexName="IDX_ORG_ATTR_NAME_VALUE"/>
             </not>
         </preConditions>
-        <createIndex indexName="IDX_ORG_ATTR_NAME_VALUE" tableName="ORGANIZATION_ATTRIBUTE">
-            <column name="NAME"/>
-            <column name="VALUE(100)"/>
-        </createIndex>
+        <sql>
+            CREATE INDEX IDX_ORG_ATTR_NAME_VALUE
+                ON ORGANIZATION_ATTRIBUTE (NAME, VALUE(100))
+        </sql>
     </changeSet>
 
     <changeSet author="phasetwo" id="add-organization-attribute-name-value-index">

--- a/src/main/resources/META-INF/jpa-changelog-phasetwo-20260421.xml
+++ b/src/main/resources/META-INF/jpa-changelog-phasetwo-20260421.xml
@@ -1,6 +1,18 @@
 <?xml version="1.1" encoding="UTF-8" standalone="no"?>
 <databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
 
+    <changeSet author="aespunto" id="add-organization-attribute-name-value-index-mariadb" dbms="mariadb">
+        <preConditions onFail="MARK_RAN">
+            <not>
+                <indexExists indexName="IDX_ORG_ATTR_NAME_VALUE"/>
+            </not>
+        </preConditions>
+        <createIndex indexName="IDX_ORG_ATTR_NAME_VALUE" tableName="ORGANIZATION_ATTRIBUTE">
+            <column name="NAME"/>
+            <column name="VALUE(100)"/>
+        </createIndex>
+    </changeSet>
+
     <changeSet author="phasetwo" id="add-organization-attribute-name-value-index">
         <preConditions onFail="MARK_RAN">
             <not>

